File: packages/skill-sdk-typescript/src/permissions.ts
const permissionPattern = /^[a-z][a-z0-9_]*(\.[a-z][a-z0-9_]*){2,5}$/u;

export function validatePermissions(values: Iterable<string>): ReadonlySet<string> {
  const permissions = new Set(values);
  for (const permission of permissions) {
    if (!permissionPattern.test(permission) || permission.includes("*")) {
      throw new Error(`Invalid or overbroad Skill permission: ${permission}`);
    }
  }
  return permissions;
}

export function effectivePermissions(
  caller: Iterable<string>,
  installation: Iterable<string>,
  manifest: Iterable<string>,
  runtime: Iterable<string>
): ReadonlySet<string> {
  const layers = [caller, installation, manifest, runtime].map(validatePermissions);
  return new Set([...layers[0]].filter((value) => layers.slice(1).every((layer) => layer.has(value))));
}

export function requirePermission(contextPermissions: ReadonlySet<string>, permission: string): void {
  validatePermissions([permission]);
  if (!contextPermissions.has(permission)) throw new Error("SKILL_PERMISSION_DENIED");
}

File: packages/skill-sdk-python/src/vav_skill_sdk/permissions.py
"""Deny-by-default permission calculations shared by SDK and runtime."""

import re
from collections.abc import Iterable

PERMISSION_PATTERN = re.compile(r"^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*){2,5}$")
FORBIDDEN_WILDCARDS = {"*", "admin.*", "secret.*", "database.*", "all_user_data.*"}


def validate_permissions(permissions: Iterable[str]) -> frozenset[str]:
    normalized = frozenset(permissions)
    invalid = sorted(
        permission
        for permission in normalized
        if permission in FORBIDDEN_WILDCARDS
        or "*" in permission
        or PERMISSION_PATTERN.fullmatch(permission) is None
    )
    if invalid:
        raise ValueError(f"unregistered or overbroad permission identifiers: {invalid}")
    return normalized


def effective_permissions(
    caller_grants: Iterable[str],
    installation_grants: Iterable[str],
    manifest_requests: Iterable[str],
    runtime_policy: Iterable[str],
) -> frozenset[str]:
    """Authority only narrows as it crosses each boundary."""
    layers = tuple(
        validate_permissions(layer)
        for layer in (caller_grants, installation_grants, manifest_requests, runtime_policy)
    )
    return frozenset.intersection(*layers)
